cpp/src/io/csv/csv_gpu.cu

```diff
@@ -570,6 +571,315 @@ __global__ void convertCsvToGdf(const char *raw_csv,
   }
 }
 
+/**
+ * @brief merge two packed row contexts (each corresponding to a block of characters)
+ * and returns the packed row context corresponding to the merged character block
+ **/
+inline __device__ uint64_t merge_row_contexts(uint64_t first_ctx, uint64_t second_ctx)
```

Comment:
AFAICT, CUDA supports bit fields. IMO the code would be _much_ more readable if we used this instead of plain uint32/64.

Reply:
It also would be horribly inefficient

Reply:
I don't see why it would have to be. Is this a known issue with CUDA and bit-fields?

Reply:
The struct would first have to be an indexable array of arbitrary bit-length values. I'd encourage you to experiment over at Godbolt.org rather than rely on assumptions.

Reply:
> The struct would first have to be an indexable array of arbitrary bit-length values.

I don't think I understand this. The struct could just have separate fields.

I assume that the language features are usable, it's true that sometimes it's not the case :)
I wrote up some sample code that's fairly similar to the row contexts: https://godbolt.org/z/UfwjCs
The generated code looks similar between bit-fields and manual packing. The hand-crafted version does seem slightly smaller.

Reply:
Heh. You don't want to use a signed type for bitfields :) Half the instructions generated will be to sign-extend & mask stuff.
Regarding the indexable array is because the 2-bit context id is indexed by a variable in many places (corresponds to a shift by a variable), so one operation for example would be:
`ctxid= ctxid_out[ctxid_in]` where ctxid_out is an array of 2-bit values: I'm not sure it's even possible to declare such a thing in any high-level SW language (not a problem for HW languages), but here it's just `ctxid = (packed_ctx >> (8 + ctxid_in * 2)) & 3;` You could avoid the constant `8` in a struct, but not the shift afaik.
In most cases, the merge operations here will result in very few instructions (which is critical for perf here).

Reply:
> You don't want to use a signed type for bitfields :)
Yup, switching to `unsigned char` shaves off two instructions from `f1()`. Focused on field sizes and forgot to about the sign. But, indexing issues aside, this sample suggests that the bit-fields are as fast as manual packing/unpacking

I haven't dug into the code enough to see the "indexing" used. Shame, this would remove so much noise in the code.

Reply:
Indexing is used in the simplest `get_row_context()` (the short-context version is inlined inside `merge_short_context` because it's only needed in one place)
The difference here is like an order-of-magnitude. With structs of bitfields, the best I could do on godbolt for the simplest `get_row_context` is a switch (and the merge calls this function 3 times per merge), which resulted in literally 5x more instructions and 3 data-dependent branches (big no-no for both cpu and gpu).
It also resulted in a *lot* more source code fwiw.

Reply:
Interesting. Do you have this saved?

Reply:
I was curious, in case compilers got a lot better since the last time I checked: [Godbolt says "Nope"](https://godbolt.org/z/kfJG95)
Note that things tend to get worse on device code, since GPU doesn't have a lot of fancy x86 zero or sign-extension on load or partial register access -> you generally don't want to pass chars around everywhere like we do (on the other hand, ptxas does some very surprisingly clever stuff with the LOP3 instruction for series of logical operations - I still have to wrap my head around that instruction)